lms/security.py

```diff
@@ -142,30 +174,30 @@ def _get_lti_user(request) -> LTIUser:
     If the request doesn't contain either valid LTI launch params or a valid
     bearer token then return ``None``.
 
+    :param: from_identity Use the lt_user from request.identity.
+        Defaults to false to avoid a circular dependency and allow
+        security policies to use this function.
+
     :rtype: models.LTIUser
     """
-    bearer_token_schema = BearerTokenSchema(request)
-
-    schemas = [
-        LTI11AuthSchema(request).lti_user,
-        partial(bearer_token_schema.lti_user, location="headers"),
-        partial(bearer_token_schema.lti_user, location="querystring"),
-        partial(bearer_token_schema.lti_user, location="form"),
-        OAuthCallbackSchema(request).lti_user,
-    ]
-    # Avoid checking for the LTI1.3 JWT if not there
-    if "id_token" in request.params:
-        # Don't replace all auth methods in case somehow we have a bogus token
-        # but try this one first.
-        schemas.insert(0, LTI13AuthSchema(request).lti_user)
-
     lti_user = None
-    for schema in schemas:
-        try:
-            lti_user = schema()
-            break
-        except ValidationError:
-            continue
+    if from_identity and request.identity and request.identity.lti_user:
+        lti_user = request.identity.lti_user
+    else:
+
+        bearer_token_schema = BearerTokenSchema(request)
+        schemas = [
+            partial(bearer_token_schema.lti_user, location="headers"),
+            partial(bearer_token_schema.lti_user, location="querystring"),
+            partial(bearer_token_schema.lti_user, location="form"),
+            OAuthCallbackSchema(request).lti_user,
```

Comment:
Plan now is to move this out one by one and then this method becomes just: `return request.identity.lti_user`

(Or get's removed and we remove every `request.lti_user` usage by request.identity.lti_user)

Reply:
:+1:

The two cases in one method is a bit ugly right now, it'll be good once these have been split out